### PR TITLE
Add support for using buffer device addresses

### PR DESCRIPTION
**Pull request by callumfare**
This enables use of physical addressing for global memory.

Requires the relevant clspv change: https://github.com/google/clspv/pull/954

I wasn't sure whether we want this to be enabled by default at this stage, so for now it requires `CLVK_PHYSICAL_ADDRESSING=1` to enable

This contribution is being made by Codeplay on behalf of Samsung.

### CODE REVIEW

**Comment by kpet** [comment]
@callumfare I had spotted but I wanted to spend a bit more time reviewing and testing that one and ran out of time last weekend :). One thing I spotted on the first read though: I think I'd want us to have two config options, one for the number of address bits and another one for physical addressing. I expect physical/64 to become the default soon but it'd be good to retain the ability to quickly test other configurations.

**Review by kpet — CHANGES_REQUESTED** [review]
Thanks for this! Nothing too surprising overall. I've compared it against what I'd done as part of my SVM/USM prototype and it's missing a few things (i.e. usage flags) but nothing I can't land myself after I rebase (this will require changes to make usage flag determination dynamic which I've needed for 1D buffer image support as well).

I've run the CTS and it turned up a good number of regressions, we still have some work to do before we're able to make physical addressing the default. I've created https://github.com/kpet/clvk/issues/477 to track the transition.

More worryingly, this change doesn't pass clvk's `api_tests` when physical addressing is enabled and I think that ought to be fixed and regression testing put in place before we land this PR. I suggest copying the test step at https://github.com/kpet/clvk/blob/main/.github/workflows/presubmit.yml#L177 into a new `API tests (physical addressing)` step. It should just require a copy/paste and a couple additional environment variables (and fixing whatever prevents the `WithCommandQueue.PodUBO` test from passing of course :)).

**Comment by kpet** [comment]
@callumfare Thanks! A few possible options:
1. Add support for Int64 to Swiftshader
2. Add support for `uint2` pointers to clspv (I remember @alan-baker was keen on seeing this done too)
3. We could land this PR with the new CI step disabled to get some forward progress (and add a task to https://github.com/kpet/clvk/issues/477). Not ideal but workable if sufficiently short-lived.

I have a strong preference for 1 or 2 but if you need 3, that's an acceptable compromise.

**Comment by callumfare** [comment]
@kpet 

I think we'd prefer to go for option 3 as we ideally want all of our pending changes to be done in the next week or so - we (Codeplay) don't have plans to continue clvk/clspv contributions after this month. If it helps, I think we're happy to leave the option disabled by default, but in the short-term enable it on a per-device basis.

File: src/kernel.hpp [hunk 1]
```diff
@@ -253,7 +253,27 @@ struct cvk_kernel_argument_values {
...
+                auto mem_downcast = icd_downcast(mem);
+                auto buff = reinterpret_cast<const cvk_buffer*>(mem_downcast);
+
+                VkBufferDeviceAddressInfo info{};
+                info.buffer = buff->vulkan_buffer();
+                info.sType = VK_STRUCTURE_TYPE_BUFFER_DEVICE_ADDRESS_INFO;
+                info.pNext = NULL;
+
+                auto dev_addr = vkGetBufferDeviceAddress(
+                    mem_downcast->context()->device()->vulkan_device(), &info);
```
[the remark this answers is not in the document]
Comment (kpet) [reply src/kernel.hpp]: Yes, that makes sense.

**Review by kpet — APPROVED** [review]

**Comment by kpet** [comment]
@callumfare Works for me. Noted, I'll try to get PRs in as soon as I can.

**Pull request merged**